### PR TITLE
Print capability

### PR DESCRIPTION
**Pull request by ZsoltBalvanyos**

### CODE REVIEW

File: src/main/scala/com/ovoenergy/comms/composer/rendering/PrintRendering.scala [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import com.ovoenergy.comms.composer.print.RenderedPrint
+import com.ovoenergy.comms.model
+import com.ovoenergy.comms.model.{CommManifest, CustomerAddress, CustomerProfile, TemplateData}
+import com.ovoenergy.comms.templates.model.template.processed.print.PrintTemplate
+
+import scala.collection.JavaConverters._
+
+object PrintRendering extends Rendering {
+
+  def renderPrint(clock: Clock)(commManifest: CommManifest,
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/composer/rendering/PrintRendering.scala]: change to renderPrintHtml

File: src/main/scala/com/ovoenergy/comms/composer/print/RenderedPrint.scala [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package com.ovoenergy.comms.composer.print
+
+case class RenderedPrint(htmlFooter: Option[String], htmlBody: String, htmlHeader: Option[String])
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/composer/print/RenderedPrint.scala]: change to RenderedPrintHtml

File: src/main/scala/com/ovoenergy/comms/composer/rendering/Rendering.scala [hunk 3]
```diff
@@ -97,23 +26,37 @@ object Rendering extends Logging {
...
       case (key, templateData) => key -> extractValueFromTemplateData(templateData)
     }
 
-    (dataAsStrings +
-      ("profile" -> profileToMap(customerProfile)) +
-      ("recipient" -> recipient.asJava) +
-      ("system" -> systemVariables(clock))).asJava
+    val a = combineJMaps(dataAsStrings.asJava, customerData)
+    val b = systemVariables(clock)
+    println("System variables: " + b)
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/composer/rendering/Rendering.scala]: Remove printlns and unnecessary variables

File: src/main/scala/com/ovoenergy/comms/composer/rendering/Rendering.scala [hunk 4]
```diff
@@ -123,27 +66,24 @@ object Rendering extends Logging {
...
-          .map {
-            case (sym, value) => (sym.name, value)
-          }
-          .toMap
-          .asJava
-      }
-      .getOrElse(Map().asJava)
-  }
 
+  def valueToMap[E, L <: HList, F <: HList](instanceToConvert: E)(
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/composer/rendering/Rendering.scala]: 😎

File: src/main/scala/com/ovoenergy/comms/composer/rendering/PrintRendering.scala [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      .getOrElse(Map.empty)
+
+    val customerAddressMap: Map[String, AnyRef] = Map("address" -> valueToMap(customerAddress).asJava)
+
+    val context = buildHandlebarsContext(
+      data,
+      customerAddressMap.asJava.combineWith(customerProfileMap.asJava),
+      clock
+    )
+
```
Comment (cb372) [inline src/main/scala/com/ovoenergy/comms/composer/rendering/PrintRendering.scala]: It's a minor detail but I would have kept the customer data typed as `Map[String, Map[String, String]]` (maybe with a nice type alias) for as long as possible, i.e. try and contain the `JMap[String, AnyRef]` nastiness as much as possible.

e.g.

```scala
val customerProfileMap: Map[String, Map[String, String]] = customerProfile
  .map(profile => Map("profile" -> valueToMap(profile)))
  .getOrElse(Map.empty)

val customerAddressMap: Map[String, Map[String, String]] = 
  Map("address" -> valueToMap(customerAddress))

val customerData: Map[String, Map[String, String]] = 
  Monoid.combine(customerProfileMap, customerAddressMap)

val context = buildHandlebarsContext(
   data,
   customerData,
   clock
 )
```

Then inside `buildHandlebarsContext` you do the work to turn nested Scala maps into nested Java maps:

```scala
customerData.mapValues(_.asJava).asJava
```

File: src/main/scala/com/ovoenergy/comms/composer/Main.scala [hunk 6]
```diff
@@ -74,7 +98,8 @@ object Main extends App {
 
   Seq(
     (emailGraph, "Email Composition"),
-    (smsGraph, "SMS Composition")
+    (smsGraph, "SMS Composition"),
+    (printGraph, "Print Composition")
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/composer/Main.scala]: I wouldn't wire up the print graph just yet, these changes can sit until we've fully implemented the PDF rendering

**Pull request merged**